quaint/src/connector/postgres/native/cache.rs

```diff
@@ -0,0 +1,394 @@
+    }
+
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error> {
+        match self.cache.get(sql, types).await {
+            Some(statement) => Ok(statement),
+            None => {
+                let stmt = client.prepare_typed(sql, types).await?;
+                self.cache.insert(sql, types, stmt.clone()).await;
+                Ok(stmt)
+            }
+        }
+    }
+}
+
+impl From<CacheSettings> for PreparedStatementLruCache {
+    fn from(settings: CacheSettings) -> Self {
+        Self::with_capacity(settings.capacity)
+    }
+}
+
+/// An LRU cache that creates and stores query type information rather than prepared statements.
+/// Queries are identified by their content with tracing information removed (which makes it
+/// possible to cache traced queries at all) and returned as instances of [`TypedQuery`]. The
+/// caching behavior is implemented in [`get_query`](Self::get_query), while statements returned
+/// from [`get_statement`](Self::get_statement) are always freshly prepared, because statements
+/// cannot be re-used when tracing information is present.
+#[derive(Debug)]
+pub struct TracingLruCache {
+    cache: InnerLruCache<Arc<QueryMetadata>>,
+}
+
+impl TracingLruCache {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            cache: InnerLruCache::with_capacity(capacity),
+        }
+    }
+}
+
+#[async_trait]
+impl QueryCache for TracingLruCache {
+    type Query<'a> = TypedQuery<'a>;
+
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<TypedQuery<'a>, Error> {
+        let sql_without_traceparent = strip_query_traceparent(sql);
+
+        let metadata = match self.cache.get(sql_without_traceparent, types).await {
+            Some(metadata) => metadata,
+            None => {
+                let stmt = client.prepare_typed(sql_without_traceparent, types).await?;
+                let metdata = Arc::new(QueryMetadata::from(&stmt));
+                self.cache.insert(sql_without_traceparent, types, metdata.clone()).await;
+                metdata
+            }
+        };
+        Ok(TypedQuery::from_sql_and_metadata(sql, metadata))
+    }
+
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error> {
+        client.prepare_typed(sql, types).await
+    }
+}
+
+impl From<CacheSettings> for TracingLruCache {
+    fn from(settings: CacheSettings) -> Self {
+        Self::with_capacity(settings.capacity)
+    }
+}
+
+/// Settings related to query caching.
+#[derive(Debug)]
+pub struct CacheSettings {
+    pub capacity: usize,
+}
+
+/// Key uniquely representing an SQL statement in the prepared statements cache.
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct QueryKey {
+    /// Hash of a string with SQL query.
+    sql: u64,
+    /// Combined hash of types for all parameters from the query.
+    types_hash: u64,
```

Comment:
I don't know much about the underlying math, but is taking a hash of a hash safe in terms of it statistical properties? Should we have a no-op hasher that just combines the inputs for the hash map in the `LruCache` itself?

Reply:
(I guess it's probably is and it should be fine, we're just doing slightly more computations than technically necessary)